src/watcher.lisp

```diff
@@ -112,8 +119,9 @@
 (defun stop-hierarchy-watcher-thread ()
   (let ((thread (with-hierarchies-lock *watcher-thread*)))
     (when thread
-      (with-logged-problems '(stop-hierarchy-watcher-thread :destroy-thread)
-        (bt:release-lock *stop-semaphore*))
+      (with-logged-problems '(stop-hierarchy-watcher-thread :stop-thread)
+        #-ecl (bt:condition-notify *stop-cv*)
```

Comment:
I was under the impression that one has to hold the associated lock when doing ``condition-notify``. However, the documentation says no such thing: https://trac.common-lisp.net/bordeaux-threads/wiki/ApiDocumentation#condition-waitcondition-variablelock.

This seems like a bt bug since the ``sb-thread:condition-notify`` documentation says

> IMPORTANT: The same mutex that is used in the corresponding CONDITION-WAIT must be held by this thread during this call.

I have no idea how this could be fixed in bt, though, since the lock is not passed to ``condition-notify`` and lots of code like the above probably already exists.

Reply:
so condition-variables are worse then lock+timeout given this non-conformity?

Reply:
or maybe we should acquire and release the lock like in ECL case, then thread affinity will be preserved and lock will be held all the time until notification is made.

Reply:
I still think a condition variable is the right solution and somehow fixing the issue in bt should be a priority. And, just for good measure, there is also sionescu/bordeaux-threads/issues/35.

I recommend just holding the lock around ``condition-notify`` since bt does not forbid that.

Maybe we should ask @sionescu about this, before deciding on anything?

Reply:
I think that wrapping it in `with-lock-held` is right thing to do. Note, that it is the same as when you use condition-notify in SBCL – it is user responsibility to acquire the lock. BT is a simple mapping for underlying implementations. I'd merge it as is.

Reply:
Yes, that's why I suggested it. It may still not be right, though (sorry for not noticing earlier). I'll make a separate comment for the issue.

The thing I'm concerned with is that ``bt:condition-notify`` does not mention anything about holding the lock in its documentation string. But that shouldn't stop this change in log4cl.